src/Mapper.cpp

```diff
@@ -1346,3 +1346,206 @@ map<int, CGRANode*>* Mapper::getReorderPath(map<CGRANode*, int>* t_path) {
+      else jsonFile<<"             \""<<tile->getID()<<"\":"<<level<<endl;
+    }
+  }
+  jsonFile<<"     }"<<endl;
+
+  jsonFile<<"}"<<endl;
+}
+
+// Reads from the referenced mapping results json file and generates variables for incremental mapping
+int Mapper::readRefMapRes(CGRA* t_cgra, DFG* t_dfg){
+  ifstream refFile("./increMapInput.json");
+  if (!refFile.good()) {
+    cout<<"Incremental mapping requires increMapInput.json in current directory!"<<endl;
+    return -1;
+  }
+  json refs;
+  refFile >> refs;
+  CGRANodeID2Level.clear();
+  for (list<DFGNode*>::iterator dfgNode=t_dfg->nodes.begin(); dfgNode != t_dfg->nodes.end(); ++dfgNode) {
+    int dfgNodeID = (*dfgNode)->getID();
+    int x = refs["Opt2TileXY"][to_string(dfgNodeID)]["x"];
+    int y = refs["Opt2TileXY"][to_string(dfgNodeID)]["y"];
+    refMapRes[*dfgNode] = t_cgra->nodes[y][x];
+    int cgraNodeID = t_cgra->nodes[y][x]->getID();
+    CGRANodeID2Level[cgraNodeID] = refs["Tile2Level"][to_string(cgraNodeID)];
+  }
+
+  return 0;
+}
+
+// Generates variables for incremental mapping
+void Mapper::sortAllocTilesByLevel(CGRA* t_cgra){
+  map<int, vector<CGRANode*>> FanIO_CGRANodes;
+  vector<int> FanIOs;
+  int numTiles = 0;
+  for (int i=0; i<t_cgra->getRows(); ++i) {
+    for (int j=0; j<t_cgra->getColumns(); ++j) {
+
+      // Records the number of FanIO for each tile. 
+      CGRANode* currentCGRANode = t_cgra->nodes[i][j];
+      if (currentCGRANode->isDisabled()) continue; // only record tiles that have DFGNodes mapped
+      int FanIO = max(currentCGRANode->getInLinks()->size(), currentCGRANode->getOutLinks()->size());
+      FanIO_CGRANodes[FanIO].push_back(currentCGRANode);
+
+      // Records FanIO in the list if it appears for the first time
+      if (find(FanIOs.begin(), FanIOs.end(), FanIO) == FanIOs.end()) {
+        FanIOs.push_back(FanIO);
+      }
+
+      numTiles++;
+    }
+  }
+  std::sort(FanIOs.rbegin(), FanIOs.rend()); // Sorts FanIOs from big to small to automatically form the level
+
+  CGRANodes_sortedByLevel.clear();
+  for (int level = 0; level < FanIOs.size(); level++) {
+    int FanIO = FanIOs[level];
+    vector<CGRANode*> tiles = FanIO_CGRANodes[FanIO];
+    CGRANodes_sortedByLevel.push_back(tiles);
+  }
+}
+
+// Generates the placement recommendation list for current DFGNode by referencing its placement in the former mapping results
+// Two principles: RPT & MBO
+list<CGRANode*> Mapper::placementGen(CGRA* t_cgra,  DFGNode* t_dfgNode){
+  list<CGRANode*> placementRecommList;
+  CGRANode* refCGRANode = refMapRes[t_dfgNode];
+  list<DFGNode*>* predNodes = t_dfgNode->getPredNodes();
+  int refLevel = CGRANodeID2Level[refCGRANode->getID()];
+  int level = refLevel;
+  int maxLevel = CGRANodes_sortedByLevel.size() - 1;
+  cout<<t_dfgNode->getOpcodeName()<<t_dfgNode->getID()<<" is mapped to Tile "<<refCGRANode->getID()<<" in the referenced mapping results, refLevel="<<refLevel<<endl;
+
+  int initLevel = level;
```

Comment:
If `level` indicates the number of FanIO, why not just `fanIO` instead of `level`?

BTW, can we replace `fanIO` with `inOutDegree`? WDYT?

Reply:
Yeah these two variables are easily confused, I use the following examples to explain the difference:

map<int, vector<CGRANode*>> FanIO_CGRANodes = {
{4, {CGRANode5, CGRANode6, CGRANode9, CGRANode10}},
{3, {CGRANode1, CGRANode2, CGRANode4, CGRANode7, CGRANode8, CGRANode11, CGRANode13, CGRANode14}},
{2, {CGRANode0, CGRANode3, CGRANode12, CGRANode15}}
}
{CGRANode5, CGRANode6, CGRANode9, CGRANode10} all have `FanIO=5` and `Level=0`
{CGRANode1, CGRANode2, CGRANode4, CGRANode7, CGRANode8, CGRANode11, CGRANode13, CGRANode14} all have `FanIO=4` and `Level=1`
{CGRANode0, CGRANode3, CGRANode12, CGRANode15} all have `FanIO=3` and `Level=2`

For an athelete in a championship, `FanIO ` is his score, while `Level` is his ranking.

I thought `I/O degree` is for DFG and `FanIO` for CGRA?
<img width="325" alt="ISCA2023_MapZero" src="https://github.com/user-attachments/assets/ecafaf58-30de-4a4e-8b5d-29f3d6934c10">
Xiangyu Kong, Yi Huang, Jianfeng Zhu, Xingchen Man, Yang Liu, Chunyang Feng, Pengfei Gou, Minggui Tang, Shaojun Wei, and Leibo Liu. 2023. MapZero: Mapping for Coarse-grained Reconfigurable Architectures with Reinforcement Learning and Monte-Carlo Tree Search. In Proceedings of the 50th Annual International Symposium on Computer Architecture (ISCA '23). Association for Computing Machinery, New York, NY, USA, Article 46, 1–14. https://doi.org/10.1145/3579371.3589081

Reply:
Larger FanIO not always has higher priority, for example in the referenced mapping results, `DFGNode10` is mapped to `CGRANode4` that has `FanIO=4`, which is a middle level of FanIO because we also have CGRANodes whose `FanIO=3` and `FanIO=5` under 4x4 CGRA. 

So when incrementalMap() choose which CGRANodes within the new resource allocation decision have highest priority for `DFGNode10` to map on, those CGRANodes with the middle level of `FanIO` will be selected. And the middle level of `FanIO` may not always equal to 4, for example in the new resource allocation decision shown in the follow image, the middle level of `FanIO` is 2 (Others levels are `FanIO=1` and `FanIO=3`).
<img width="68" alt="example" src="https://github.com/user-attachments/assets/61538316-7bf0-4111-9940-502f083b6b47">
Comparing the `FanIO` may mislead incrementalMap() to select the CGRANodes with `FanIO=3` if you think larger FanIO has higher priority? 

So sorry I still think `level` may be more appropriate than `priroity`.

Reply:
How could there exist `FanIO=5` under 4x4 CGRA? I assume you are talking about `FanIO=4`.  It sounds like you are ordering the tiles based on how their FanIO is closed to the FanIO of the tile the DFGNode was originally mapped on. Can you please schedule a 20min one-on-one meeting with me?

Reply:
Plz add comment "The `level` is used to ordering the CGRANodes based on the `FanIO`. Though `FanIO` of each CGRANode would change for different CGRA architectures, the DFGNode prefers to being mapped onto the CGRANode with same `level`."